### DISCUSSION TITLE
[JS] 메모리 누수

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#78] [프론트 기술 질문 공간]
메모리 누수에 대해 얼마나 알고 계시나요?
아는 대로 설명 해주세요.

**Comment by michoball** [comment]
javascript는 가비지 컬렉터가 있는 언어라 대부분의 경우 사용되지 않고 자리를 차지하는 메모리를 정기적으로 삭제합니다.
하지만 개발자가 실수로 사용하지 않는 데이터의 위치에 대한 참조를 계속 남겨두어 해당 메모리를 삭제할 수 없게 되고 이는 메모리 누수를 만듭니다. 

메모리 누수를 만드는 케이스

1. global variables (use strict mode 가 아닐 때)
```javascript
function foo(arg) {
    bar = "this is a hidden global variable";
}
```

 다음과 같이 var, let, const 를 사용하지 않고 사용하면 window.bar 로 사용한게 된다. 

2. setTimer 나 callback  사용오류

```javascript
var someResource = getData();
setInterval(function() {
    var node = document.getElementById('Node');
    if(node) {
        // Do stuff with node and someResource.
        node.innerHTML = JSON.stringify(someResource));
    }
}, 1000);
```

위 경우 setInterval로 1초마다 안의 함수가 계속 실행되고 안의 함수에서 someResource를 계속 참조해서 사용한다. 
즉 setInterval은 계속 실행되므로 가비지컬렉터에 잡히지 않고 someResource또한 그렇다. 


3. Dom 참조 

```javascript
var elements = {
    button: document.getElementById('button'),
    image: document.getElementById('image'),
    text: document.getElementById('text')
};

function doStuff() {
    image.src = 'http://some.url/image';
    button.click();
    console.log(text.innerHTML);
    // Much more logic
}

function removeButton() {
    // The button is a direct child of body.
    document.body.removeChild(document.getElementById('button'));

    // At this point, we still have a reference to #button in the global
    // elements dictionary. In other words, the button element is still in
    // memory and cannot be collected by the GC.
}
```
위처럼 돔 요소에 대한 참조를 특정 자료구조 안에 넣어두면 다른 곳에서 특정 돔 요소를 지우더라도 참조 요소가 남아있어 메모리 누수가 날 수 있습니다. 

5. 클로저에 의한 누수

```javascript
var theThing = null;
var replaceThing = function () {
  var originalThing = theThing;
  var unused = function () {
    if (originalThing)
      console.log("hi");
  };
  theThing = {
    longStr: new Array(1000000).join('*'),
    someMethod: function () {
      console.log(someMessage);
    }
  };
};
setInterval(replaceThing, 1000);
```
위 코드에서 1초마다 replaceThing이 계속 불리면 replaceThing안의 originalThing은 그전의 theThing을 참조하고 이는 unused 함수에서 사용된다. 그리고 theThing 에 새로운 객체가 선언되는 구조다.

이러면 unused는 쓰이지 않더라고 계속 이전 originalThing에 대한 참조를 가지므로 지속적으로 이전 값에 대한 참조를 하게 되고, 
theThing안에 꽤나 큰 배열이 매번 생성된다면 엄청 큰 메모리 누수를 불러일으킨다. 

참고자료입니다.  : https://auth0.com/blog/four-types-of-leaks-in-your-javascript-code-and-how-to-get-rid-of-them/

**Comment by WooGie911** [comment]
자바스크립트는 할당한 메모리를 애플리케이션에서 여전히 사용중인지 검사해 사용하지 않는 메모리를 관리하는 가비지 컬렉트 언어입니다.

자바스크립트에서의 메모리 누수 주요 원인은 '예상치 못한 참조' 입니다.
메모리 누수는 애플리케이션 속도저하, 예기치 못한 종료, 느린 응답속도 등과 같이 많은 문제를 야기할 수 있습니다.

메모리 누수의 일반적인 형태로는
- 우발적으로 생긴 전역변수
- 잊혀진 타이머와 콜백
- DOM을 외부에서 참조
- 클로저
등이 있습니다.

마지막으로 가비지 컬렉터는 비결정성인 특징을 가지고 있습니다. 
즉, 언제 수집되는지 정확히 예측할 수 없습니다. 
그러나 대부분의 경우 메모리 할당이 이뤄지는 경우에만 수집을 진행합니다.
